File: java/src/test/java/it/giannibombelli/wsc2026/payment/application/usecases/TransactionRejectingTest.java
package it.giannibombelli.wsc2026.payment.application.usecases;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentEvent;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentResultEvents;
import it.giannibombelli.wsc2026.payment.domain.events.TransactionRejected;
import it.giannibombelli.wsc2026.payment.domain.events.TransactionStarted;
import it.giannibombelli.wsc2026.payment.domain.payment.Payment;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentStatus;
import it.giannibombelli.wsc2026.payment.domain.payment.Provider;
import it.giannibombelli.wsc2026.payment.domain.payment.ProviderReference;
import it.giannibombelli.wsc2026.payment.domain.payment.TransactionId;
import it.giannibombelli.wsc2026.payment.domain.policies.PaymentRejection;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentRepository;
import it.giannibombelli.wsc2026.payment.infrastructure.SqlitePaymentRepository;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import it.giannibombelli.wsc2026.testsupport.events.CapturingEventPublisher;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import javax.sql.DataSource;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.Optional;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class TransactionRejectingTest {
    private PaymentRepository repository;
    private CapturingEventPublisher<PaymentEvent> publisher;
    private TransactionRejecting transactionRejecting;

    @BeforeEach
    void setUp() {
        final DataSource dataSource = DatabaseSetup.initializeInMemoryDb("payment");
        repository = new SqlitePaymentRepository(dataSource);
        publisher = new CapturingEventPublisher<>();
        transactionRejecting = new TransactionRejecting(repository, publisher, new PaymentRejection());
    }

    @Nested
    class Construction {
        @Test
        void rejectNullRepository() {
            assertThatThrownBy(() -> new TransactionRejecting(null, publisher, new PaymentRejection()))
                .isInstanceOf(NullPointerException.class);
        }

        @Test
        void rejectNullPublisher() {
            assertThatThrownBy(() -> new TransactionRejecting(repository, null, new PaymentRejection()))
                .isInstanceOf(NullPointerException.class);
        }

        @Test
        void rejectNullPolicy() {
            assertThatThrownBy(() -> new TransactionRejecting(repository, publisher, null))
                .isInstanceOf(NullPointerException.class);
        }
    }

    @Nested
    class Execution {
        @Test
        void rejectNullEvent() {
            assertThatThrownBy(() -> transactionRejecting.on(null))
                .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        void shouldRejectTransaction() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            String clientReference = UUID.randomUUID().toString();
            Money amount = new Money(new BigDecimal("50.00"));
            Instant requestedAt = Instant.parse("2026-06-07T10:00:00Z");
            Payment payment = Payment.request(paymentId, new ClientReference(clientReference), amount, new Timestamp(requestedAt));
            repository.save(payment);

            TransactionId transactionId = startTransaction(paymentId, amount);

            transactionRejecting.on(new TransactionRejected(paymentId, Provider.PAYPAL, transactionId, new Description("declined")));

            Optional<Payment> loaded = repository.findById(paymentId);
            assertThat(loaded).isPresent();
            assertThat(loaded.get().status()).isEqualTo(PaymentStatus.REJECTED);
        }

        @Test
        void shouldPublishRejectedEvent() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            String clientReference = UUID.randomUUID().toString();
            Money amount = new Money(new BigDecimal("50.00"));
            Instant requestedAt = Instant.parse("2026-06-07T10:00:00Z");
            Payment payment = Payment.request(paymentId, new ClientReference(clientReference), amount, new Timestamp(requestedAt));
            repository.save(payment);

            TransactionId transactionId = startTransaction(paymentId, amount);

            transactionRejecting.on(new TransactionRejected(paymentId, Provider.PAYPAL, transactionId, new Description("declined")));

            assertThat(publisher.events()).hasSize(1);
            assertThat(publisher.events().get(0)).isInstanceOf(PaymentResultEvents.PaymentRejected.class);
            PaymentResultEvents.PaymentRejected rejected = (PaymentResultEvents.PaymentRejected) publisher.events().get(0);
            assertThat(rejected.aggregateId()).isEqualTo(paymentId);
            assertThat(rejected.clientReference().value()).isEqualTo(clientReference);
        }

        @Test
        void shouldFailIfPaymentNotFound() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            TransactionId transactionId = new TransactionId(UUID.randomUUID());

            assertThatThrownBy(() -> transactionRejecting.on(
                new TransactionRejected(paymentId, Provider.PAYPAL, transactionId, new Description("declined"))))
                .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        void shouldFailIfAlreadyAccepted() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            String clientReference = UUID.randomUUID().toString();
            Money amount = new Money(new BigDecimal("50.00"));
            Instant requestedAt = Instant.parse("2026-06-07T10:00:00Z");
            Payment payment = Payment.request(paymentId, new ClientReference(clientReference), amount, new Timestamp(requestedAt));
            TransactionId acceptedTransactionId = EntityId.generate(TransactionId::new);
            payment.startTransaction(
                acceptedTransactionId,
                Provider.PAYPAL,
                new ProviderReference(UUID.randomUUID()),
                amount,
                new Timestamp(requestedAt.plusSeconds(1))
            );
            payment.acceptTransaction(acceptedTransactionId, new Timestamp(requestedAt.plusSeconds(60)));
            repository.save(payment);

            TransactionId transactionId = new TransactionId(UUID.randomUUID());
            assertThatThrownBy(() -> transactionRejecting.on(
                new TransactionRejected(paymentId, Provider.PAYPAL, transactionId, new Description("declined"))))
                .isInstanceOf(IllegalStateException.class);
        }

        private TransactionId startTransaction(PaymentId paymentId, Money amount) {
            Payment payment = repository.findById(paymentId).orElseThrow();
            TransactionId transactionId = EntityId.generate(TransactionId::new);
            TransactionStarted started = payment.startTransaction(
                transactionId,
                Provider.PAYPAL,
                new ProviderReference(UUID.randomUUID()),
                amount,
                new Timestamp(Instant.now())
            );
            repository.save(payment);
            return started.transactionId();
        }
    }
}

File: typescript/tests/payment/application/usecases/transactionRejecting.test.ts
import { beforeEach, describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { ClientReference } from '@/common/domain/primitive/clientReference.js';
import { Description } from '@/common/domain/primitive/description.js';
import { Money } from '@/common/domain/primitive/money.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { TransactionRejecting } from '@/payment/application/usecases/transactionRejecting.js';
import { PaymentEvent } from '@/payment/domain/events/paymentEvent.js';
import type { PaymentRejected } from '@/payment/domain/events/paymentResultEvents.js';
import {
  transactionRejected,
  type TransactionRejected,
} from '@/payment/domain/events/transactionRejected.js';
import { Payment } from '@/payment/domain/payment/payment.js';
import { PaymentId } from '@/payment/domain/payment/paymentId.js';
import { PaymentStatus } from '@/payment/domain/payment/paymentStatus.js';
import { Provider } from '@/payment/domain/payment/provider.js';
import { ProviderReference } from '@/payment/domain/payment/providerReference.js';
import { TransactionId } from '@/payment/domain/payment/transactionId.js';
import { PaymentRejection } from '@/payment/domain/policies/paymentRejection.js';
import { SqlitePaymentRepository } from '@/payment/infrastructure/sqlitePaymentRepository.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import { CapturingEventPublisher } from '../../../testsupport/events/capturingEventPublisher.js';
import { DatabaseSetup } from '../../../testsupport/databaseSetup.js';

describe('TransactionRejecting', () => {
  let repository: SqlitePaymentRepository;
  let publisher: CapturingEventPublisher<PaymentEvent>;
  let transactionRejecting: TransactionRejecting;

  beforeEach(() => {
    const database = DatabaseSetup.initializeInMemoryDb('payment');
    repository = new SqlitePaymentRepository(database);
    publisher = new CapturingEventPublisher();
    transactionRejecting = new TransactionRejecting(repository, publisher, new PaymentRejection());
  });

  describe('construction', () => {
    it('rejects null repository', () => {
      expect(() => new TransactionRejecting(null as unknown as SqlitePaymentRepository, publisher, new PaymentRejection())).toThrow();
    });

    it('rejects null publisher', () => {
      expect(() => new TransactionRejecting(repository, null as unknown as CapturingEventPublisher<PaymentEvent>, new PaymentRejection())).toThrow();
    });

    it('rejects null policy', () => {
      expect(() => new TransactionRejecting(repository, publisher, null as unknown as PaymentRejection)).toThrow();
    });
  });

  describe('execution', () => {
    it('rejects null event', () => {
      expect(() => transactionRejecting.on(null as unknown as TransactionRejected)).toThrow();
    });

    it('should reject transaction', () => {
      const paymentId = seedPayment();
      const transactionId = startTransaction(paymentId, new Money(50));

      transactionRejecting.on(transactionRejected(paymentId, Provider.PAYPAL, transactionId, new Description('declined')));

      const loaded = repository.findById(paymentId);
      expect(loaded).not.toBeNull();
      expect(loaded!.status()).toEqual(PaymentStatus.REJECTED);
    });

    it('should publish rejected event', () => {
      const paymentId = seedPayment();
      const transactionId = startTransaction(paymentId, new Money(50));

      transactionRejecting.on(transactionRejected(paymentId, Provider.PAYPAL, transactionId, new Description('declined')));

      expect(publisher.events()).toHaveLength(1);
      expect(publisher.events()[0].kind).toBe('PaymentRejected');
      const rejected = publisher.events()[0] as PaymentRejected;
      expect(rejected.aggregateId).toEqual(paymentId);
    });

    it('should fail if payment not found', () => {
      const paymentId = new PaymentId(Uuid.generate());
      const transactionId = new TransactionId(Uuid.generate());

      expect(() => transactionRejecting.on(transactionRejected(paymentId, Provider.PAYPAL, transactionId, new Description('declined')))).toThrow();
    });

    it('should fail if already accepted', () => {
      const paymentId = seedPayment();
      const payment = repository.findById(paymentId)!;
      const acceptedTransactionId = generateId((value) => new TransactionId(value));
      const startedAt = new Timestamp(new Date('2026-06-07T10:01:00.000Z'));
      const acceptedAt = new Timestamp(new Date('2026-06-07T10:30:00.000Z'));
      payment.startTransaction(
        acceptedTransactionId,
        Provider.PAYPAL,
        new ProviderReference(Uuid.generate()),
        new Money(50),
        startedAt
      );
      payment.acceptTransaction(acceptedTransactionId, acceptedAt);
      repository.save(payment);

      const transactionId = new TransactionId(Uuid.generate());
      expect(() => transactionRejecting.on(transactionRejected(paymentId, Provider.PAYPAL, transactionId, new Description('declined')))).toThrow();
    });
  });

  function seedPayment(): PaymentId {
    const paymentId = generateId((value) => new PaymentId(value));
    const payment = Payment.request(
      paymentId,
      new ClientReference(Uuid.fromString(crypto.randomUUID())),
      new Money(50),
      new Timestamp(new Date('2026-06-07T10:00:00.000Z'))
    );
    repository.save(payment);
    return paymentId;
  }

  function startTransaction(paymentId: PaymentId, amount: Money): TransactionId {
    const payment = repository.findById(paymentId)!;
    const transactionId = generateId((value) => new TransactionId(value));
    const started = payment.startTransaction(
      transactionId,
      Provider.PAYPAL,
      new ProviderReference(Uuid.generate()),
      amount,
      Timestamp.now()
    );
    repository.save(payment);
    return started.transactionId;
  }
});
